### PR TITLE
Quiz for generic functions and recursions

### PR DESCRIPTION
**Pull request by julien-truffaut**

### CODE REVIEW

File: notes/03-data-processing/03b-quiz-recursions.MD [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+sealed trait MovingStorage
+case class Box(width: Int, height: Int) extends MovingStorage
+case class Truck(foot: Int) extends MovingStorage
+```
+
+1. Person
+1. (c) Order
+1. MovingStorage
+1. Option
+1. (c) java.io.File
```
Comment (nikbsf) [inline notes/03-data-processing/03b-quiz-recursions.MD:19]: This one is confusing. Despite the fact that you could work with `java.io.File` in a recursive manner, it doesn't look like a data structure at all. You wouldn't get child elements from it without performing some IO.

File: notes/03-data-processing/03b-quiz-recursions.MD [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+#### Explanation
+
+The most common mistake of recursions is to have invalid termination conditions, for example,
+if we check that `n == 0` instead of `n <= 0`. This may cause the recursion to run forever or
+to terminate earlier/later than expected. It is good practice using a unit test to verify a
+recursion terminates correctly.
+
+If a function is not tail-recursive, then it will throw a `StackOverflowError`
+if it calls itself too many times. We can ensure this doesn't happen by limiting the number of
+recursive calls (e.g. to a few hundreds) or by making the function tail-recursive (not always possible).
```
Comment (nikbsf) [inline notes/03-data-processing/03b-quiz-recursions.MD:74]: Thought of adding a separate question on which kind of algorithms could be refactored to become tail-recursive?

File: notes/03-data-processing/03b-quiz-recursions.MD (hunk 1 above)
Comment (julien-truffaut) [reply notes/03-data-processing/03b-quiz-recursions.MD:19]: That's true, a `java.io.File` is not a recursive ADT, but in the video about recursions, I spend quite a lot of time talking about the recursive nature of files. The other example I used in the lecture is JSON. Maybe less confusing.

Comment (nikbsf) [reply notes/03-data-processing/03b-quiz-recursions.MD]: Yeah, JSON would be much less confusing.

File: notes/03-data-processing/03b-quiz-recursions.MD (hunk 2 above)
Comment (julien-truffaut) [reply notes/03-data-processing/03b-quiz-recursions.MD]: I added a question that covers tail-recursion. Let me know what you think.

**Pull request merged**